### PR TITLE
Create ext_index.json

### PR DESCRIPTION
**Pull request by awestover**
# Problem
`compute_ctl` expects needs a file `ext_index.json` containing the data necessary to create control files, and containing the locations of extension archives in remote storage. 
The `ext_index.json` file should be generated by `Dockerfile.compute-node`.

# Summary of Changes

The layout of the S3 bucket is as follows:
```
5615610098 // this is an extension build number
├── v14
│   ├── extensions
│   │   ├── anon.tar.zstd
│   │   └── embedding.tar.zstd
│   └── ext_index.json
└── v15
    ├── extensions
    │   ├── anon.tar.zstd
    │   └── embedding.tar.zstd
    └── ext_index.json
5615261079
├── v14
│   ├── extensions
│   │   └── anon.tar.zstd
│   └── ext_index.json
└── v15
    ├── extensions
    │   └── anon.tar.zstd
    └── ext_index.json
5623261088
├── v14
│   ├── extensions
│   │   └── embedding.tar.zstd
│   └── ext_index.json
└── v15
    ├── extensions
    │   └── embedding.tar.zstd
    └── ext_index.json
```

`ext_index.json` stores the control files and location of extension archives.

More specifically, here is an example `ext_index.json`
```
{
  "embedding": {
    "control_data": {
      "embedding.control": "comment = 'hnsw index' \ndefault_version = '0.1.0' \nmodule_pathname = '$libdir/embedding' \nrelocatable = true \ntrusted = true"
    },
    "archive_path": "5623261088/v15/extensions/embedding.tar.zstd"
  },
  "anon": {
    "control_data": {
      "anon.control": "# PostgreSQL Anonymizer (anon) extension \ncomment = 'Data anonymization tools' \ndefault_version = '1.1.0' \ndirectory='extension/anon' \nrelocatable = false \nrequires = 'pgcrypto' \nsuperuser = false \nmodule_pathname = '$libdir/anon' \ntrusted = true \n"
    },
    "archive_path": "5615261079/v15/extensions/anon.tar.zstd"
  }
}
```

### CODE REVIEW

File: Dockerfile.compute-node [hunk 1]
```diff
@@ -773,14 +773,28 @@ RUN rm /usr/local/pgsql/lib/lib*.a
 # Extenstion only
 #
 #########################################################################################
+FROM python:3.11-slim-bookworm AS alekpython
```
Comment (lubennikovaav) [inline Dockerfile.compute-node:776]:
```suggestion
FROM python:3.11-slim-bookworm AS generate-ext-index
```

File: Dockerfile.compute-node [hunk 2]
```diff
@@ -773,14 +773,28 @@ RUN rm /usr/local/pgsql/lib/lib*.a
...
+ARG PG_VERSION
+ARG BUILD_TAG
+# define extension build numbers
+# NOTE: it is *not* necessary for the build tag to be BUILD_TAG in particular,
+# you should update the build_tag for each extension only when you build it
+# in the future it should be something like this: 5615610098
```
Comment (lubennikovaav) [inline Dockerfile.compute-node:782]:
```suggestion
```

the statement above is clear enough. referring to some number doesn't help

File: Dockerfile.compute-node [hunk 3]
```diff
@@ -773,14 +773,28 @@ RUN rm /usr/local/pgsql/lib/lib*.a
...
 FROM scratch AS postgres-extensions
 # After the transition this layer will include all extensitons.
 # As for now, it's only a couple for testing purposses
-#
-# # Default extensions
-COPY --from=kq-imcx-pg-build /extensions/ /extensions/
-COPY --from=pg-anon-pg-build /extensions/ /extensions/
-COPY --from=postgis-build /extensions/ /extensions/
+COPY --from=alekpython /extensions/ /extensions/
+COPY --from=alekpython /ext_index.json /ext_index.json
```
Comment (lubennikovaav) [inline Dockerfile.compute-node:797]:
```suggestion
COPY --from=generate-ext-index /extensions/ /extensions/
COPY --from=generate-ext-index /ext_index.json /ext_index.json
```

File: scripts/combine_control_files.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
+import json
```
Comment (lubennikovaav) [inline scripts/combine_control_files.py:1]:
```suggestion
# Script to generate ext_index.json metadata file
# that stores content of the control files and location of extension archives
#  for all extensions in /extensions subdir.
import json
```

File: scripts/combine_control_files.py [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import os
+import sys
+from os.path import isdir, join
+
+# import shutil
+
+pg_version = sys.argv[1]
+#  build_tag = sys.argv[2]
```
Comment (lubennikovaav) [inline scripts/combine_control_files.py:9]: Why don't we use build_tag?
Should it be uncommented?

File: Dockerfile.compute-node [hunk 6]
```diff
@@ -773,14 +773,28 @@ RUN rm /usr/local/pgsql/lib/lib*.a
...
+FROM python:3.11-slim-bookworm AS alekpython
+ARG PG_VERSION
+ARG BUILD_TAG
+# define extension build numbers
+# NOTE: it is *not* necessary for the build tag to be BUILD_TAG in particular,
+# you should update the build_tag for each extension only when you build it
+# in the future it should be something like this: 5615610098
+RUN echo "kq_imcx ${BUILD_TAG}" >>  build_numbers.txt
+RUN echo "anon ${BUILD_TAG}" >>  build_numbers.txt
+# RUN echo "postgis-build 5615610098" >>  build_numbers.txt
```
Comment (lubennikovaav) [inline Dockerfile.compute-node:785]: Please add a comment about `build_numbers.txt` purpose.

File: scripts/combine_control_files.py (hunk 5 above)
Comment (awestover) [reply scripts/combine_control_files.py:9]: we only need it once we do de-duplication #4789

File: Dockerfile.compute-node (hunk 1 above)
Comment (awestover) [reply Dockerfile.compute-node:776]: changed

File: Dockerfile.compute-node (hunk 3 above)
Comment (awestover) [reply Dockerfile.compute-node:797]: yes

File: scripts/combine_control_files.py (hunk 4 above)
Comment (awestover) [reply scripts/combine_control_files.py:1]: added

File: Dockerfile.compute-node (hunk 6 above)
Comment (awestover) [reply Dockerfile.compute-node:785]: done

File: scripts/combine_control_files.py (hunk 5 above)
Comment (lubennikovaav) [reply scripts/combine_control_files.py:9]: got it.
please, add a comment about that

Comment (awestover) [reply scripts/combine_control_files.py]: I'm just deleting the code for now, put it in #4789

File: .github/workflows/build_and_test.yml [hunk 7]
```diff
@@ -977,13 +977,13 @@ jobs:
           mkdir -p extensions-to-upload
 
           docker cp ${{ steps.create-container.outputs.EID }}:/extensions/ ./extensions-to-upload/
+          docker cp ${{ steps.create-container.outputs.EID }}:/ext_index.json ./extensions-to-upload/
 
           # Remove control files, they'll be replaced by an index file
-          rm -f ./extensions-to-upload/*.control
+          # TODO: this doesn't actually work because control files are in subdirectories
+          rm -f ./extensions-to-upload/extensions/*.control
```
Comment (lubennikovaav) [inline .github/workflows/build_and_test.yml:984]: let's fix this?

this should work:
`find  ./extensions-to-upload/extensions/ -name '*.control' | xargs rm`

Comment (awestover) [reply .github/workflows/build_and_test.yml:984]: yes let's get rid of these files

**Comment by awestover** [comment]
ok there is yet another problem with the file names 
fixing it now

**Comment by lubennikovaav** [comment]
closed in favor of #4792

**Pull request closed without merging**